// angular2-apollo/build/src/Angular2Apollo.js
import { Injectable } from '@angular/core';
import { rxify } from 'apollo-client-rxjs';
import { ApolloClient } from 'apollo-client';
import { from } from 'rxjs/observable/from';
import { fromPromise } from 'rxjs/observable/fromPromise';
import { ApolloQueryObservable } from './ApolloQueryObservable';
export var Angular2Apollo = (function () {
    /**
     * @param {?} client
     */
    function Angular2Apollo(client) {
        this.client = client;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    Angular2Apollo.prototype.watchQuery = function (options) {
        return new ApolloQueryObservable(rxify(this.client.watchQuery)(options));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Angular2Apollo.prototype.query = function (options) {
        return fromPromise(this.client.query(options));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Angular2Apollo.prototype.mutate = function (options) {
        return fromPromise(this.client.mutate(options));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    Angular2Apollo.prototype.subscribe = function (options) {
        return from(this.client.subscribe(options));
    };
    /**
     * @return {?}
     */
    Angular2Apollo.prototype.getClient = function () {
        return this.client;
    };
    Angular2Apollo.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    Angular2Apollo.ctorParameters = function () { return [
        { type: ApolloClient, },
    ]; };
    return Angular2Apollo;
}());
function Angular2Apollo_tsickle_Closure_declarations() {
    /** @type {?} */
    Angular2Apollo.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    Angular2Apollo.ctorParameters;
    /** @type {?} */
    Angular2Apollo.prototype.client;
}
