// angular2-apollo/build/src/ApolloModule.js
import { NgModule } from '@angular/core';
import { Angular2Apollo } from './Angular2Apollo';
import { SelectPipe } from './SelectPipe';
import { APOLLO_CLIENT_WRAPPER, APOLLO_CLIENT_INSTANCE } from './tokens';
export var /** @type {?} */ APOLLO_DIRECTIVES = [
    SelectPipe,
];
export var /** @type {?} */ APOLLO_PROVIDERS = [
    provideAngular2Apollo(),
];
/**
 * @return {?}
 */
export function provideAngular2Apollo() {
    return {
        provide: Angular2Apollo,
        useFactory: createAngular2Apollo,
        deps: [APOLLO_CLIENT_INSTANCE],
    };
}
/**
 * @param {?} client
 * @return {?}
 */
export function createAngular2Apollo(client) {
    return new Angular2Apollo(client);
}
/**
 * @param {?} clientFn
 * @return {?}
 */
export function getApolloClient(clientFn) {
    return clientFn();
}
/**
 * @param {?} clientFn
 * @return {?}
 */
export function defaultApolloClient(clientFn) {
    return [{
            provide: APOLLO_CLIENT_WRAPPER,
            useValue: clientFn,
        }, {
            provide: APOLLO_CLIENT_INSTANCE,
            useFactory: getApolloClient,
            deps: [APOLLO_CLIENT_WRAPPER],
        }];
}
export var ApolloModule = (function () {
    function ApolloModule() {
    }
    /**
     * @param {?} clientFn
     * @return {?}
     */
    ApolloModule.withClient = function (clientFn) {
        return {
            ngModule: ApolloModule,
            providers: [
                APOLLO_PROVIDERS,
                defaultApolloClient(clientFn),
            ],
        };
    };
    ApolloModule.decorators = [
        { type: NgModule, args: [{
                    declarations: APOLLO_DIRECTIVES,
                    exports: APOLLO_DIRECTIVES,
                },] },
    ];
    /** @nocollapse */
    ApolloModule.ctorParameters = function () { return []; };
    return ApolloModule;
}());
function ApolloModule_tsickle_Closure_declarations() {
    /** @type {?} */
    ApolloModule.decorators;
    /**
     * @nocollapse
     * @type {?}
     */
    ApolloModule.ctorParameters;
}
