// apollo-angular/build/bundles/apollo.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('apollo-client-rxjs'), require('apollo-client'), require('rxjs/observable/from'), require('rxjs/observable/fromPromise')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', 'apollo-client-rxjs', 'apollo-client', 'rxjs/observable/from', 'rxjs/observable/fromPromise'], factory) :
	(factory((global.ng = global.ng || {}, global.ng.apollo = global.ng.apollo || {}),global.ng.core,global.apollo.rxjs,global.apollo,global.Rx.Observable,global.Rx.Observable));
}(this, (function (exports,_angular_core,apolloClientRxjs,apolloClient,rxjs_observable_from,rxjs_observable_fromPromise) { 'use strict';

var SelectPipe = (function () {
    function SelectPipe() {
    }
    /**
     * @param {?} obj
     * @param {?=} name
     * @return {?}
     */
    SelectPipe.prototype.transform = function (obj, name) {
        if (name === void 0) { name = ''; }
        if (obj && name !== '') {
            // for Apollo decorator
            if (obj[name]) {
                return obj[name];
            }
            // for Angular2Apollo.watchQuery
            if (obj.data && obj.data[name]) {
                return obj.data[name];
            }
        }
    };
    SelectPipe.decorators = [
        { type: _angular_core.Pipe, args: [{
                    name: 'select',
                },] },
    ];
    /** @nocollapse */
    SelectPipe.ctorParameters = function () { return []; };
    return SelectPipe;
}());

var __extends$1 = (undefined && undefined.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var ApolloQueryObservable = (function (_super) {
    __extends$1(ApolloQueryObservable, _super);
    function ApolloQueryObservable() {
        _super.apply(this, arguments);
    }
    return ApolloQueryObservable;
}(apolloClientRxjs.RxObservableQuery));

var CLIENT_MAP_WRAPPER = new _angular_core.OpaqueToken('apollo/client-map-wrapper');
var CLIENT_MAP = new _angular_core.OpaqueToken('apollo/client-map');

var __extends = (undefined && undefined.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
/**
 * Base class that handles ApolloClient
 */
var ApolloBase = (function () {
    /**
     * @param {?} client
     */
    function ApolloBase(client) {
        this.client = client;
    }
    /**
     * @param {?} options
     * @return {?}
     */
    ApolloBase.prototype.watchQuery = function (options) {
        return new ApolloQueryObservable(apolloClientRxjs.rxify(this.client.watchQuery)(options));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    ApolloBase.prototype.query = function (options) {
        return rxjs_observable_fromPromise.fromPromise(this.client.query(options));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    ApolloBase.prototype.mutate = function (options) {
        return rxjs_observable_fromPromise.fromPromise(this.client.mutate(options));
    };
    /**
     * @param {?} options
     * @return {?}
     */
    ApolloBase.prototype.subscribe = function (options) {
        return rxjs_observable_from.from(this.client.subscribe(options));
    };
    /**
     * @return {?}
     */
    ApolloBase.prototype.getClient = function () {
        return this.client;
    };
    ApolloBase.decorators = [
        { type: _angular_core.Injectable },
    ];
    /** @nocollapse */
    ApolloBase.ctorParameters = function () { return [
        { type: apolloClient.ApolloClient, },
    ]; };
    return ApolloBase;
}());
/**
 * Container service that works just like the ApolloBase but includes named ApolloClients
 */
var Apollo = (function (_super) {
    __extends(Apollo, _super);
    /**
     * @param {?} clientMap
     */
    function Apollo(clientMap) {
        _super.call(this, clientMap['default']);
        this.map = new Map();
        for (var name_1 in clientMap) {
            if (typeof name_1 === 'string' && name_1 !== 'default') {
                this.map.set(name_1, new ApolloBase(clientMap[name_1]));
            }
        }
    }
    /**
     * @return {?}
     */
    Apollo.prototype.default = function () {
        return this;
    };
    /**
     * @param {?} name
     * @return {?}
     */
    Apollo.prototype.use = function (name) {
        if (name === 'default') {
            return this.default();
        }
        return this.map.get(name);
    };
    Apollo.decorators = [
        { type: _angular_core.Injectable },
    ];
    /** @nocollapse */
    Apollo.ctorParameters = function () { return [
        null,
    ]; };
    return Apollo;
}(ApolloBase));
/**
 * Invokes a ClientMapWrapper
 * @param {?} configWrapper
 * @return {?}
 */
function getClientMap(configWrapper) {
    var /** @type {?} */ config = configWrapper();
    if (config instanceof apolloClient.ApolloClient) {
        return { default: config };
    }
    return config;
}
/**
 * Provides a value for a map and a wrapper
 * @param {?} configWrapper
 * @return {?}
 */
function provideClientMap(configWrapper) {
    return [{
            provide: CLIENT_MAP_WRAPPER,
            useValue: configWrapper,
        }, {
            provide: CLIENT_MAP,
            useFactory: getClientMap,
            deps: [CLIENT_MAP_WRAPPER],
        }];
}
/**
 * Provides the Apollo service
 */
var provideApollo = {
    provide: Apollo,
    useFactory: createApollo,
    deps: [CLIENT_MAP],
};
/**
 * @param {?} clientMap
 * @return {?}
 */
function createApollo(clientMap) {
    return new Apollo(clientMap);
}

var APOLLO_DIRECTIVES = [
    SelectPipe,
];
var APOLLO_PROVIDERS = [
    provideApollo,
];
/**
 * @param {?} clientFn
 * @return {?}
 */
function defaultApolloClient(clientFn) {
    return provideClientMap(clientFn);
}
var ApolloModule = (function () {
    function ApolloModule() {
    }
    /**
     * @param {?} clientFn
     * @return {?}
     */
    ApolloModule.withClient = function (clientFn) {
        return {
            ngModule: ApolloModule,
            providers: [
                APOLLO_PROVIDERS,
                defaultApolloClient(clientFn),
            ],
        };
    };
    /**
     * Defines a map of ApolloClients or a single instance
     * @param {?} clientMapFn
     * @return {?}
     */
    ApolloModule.forRoot = function (clientMapFn) {
        return {
            ngModule: ApolloModule,
            providers: [
                APOLLO_PROVIDERS,
                provideClientMap(clientMapFn),
            ],
        };
    };
    ApolloModule.decorators = [
        { type: _angular_core.NgModule, args: [{
                    declarations: APOLLO_DIRECTIVES,
                    exports: APOLLO_DIRECTIVES,
                },] },
    ];
    /** @nocollapse */
    ApolloModule.ctorParameters = function () { return []; };
    return ApolloModule;
}());

exports.SelectPipe = SelectPipe;
exports.Apollo = Apollo;
exports.ApolloBase = ApolloBase;
exports.provideClientMap = provideClientMap;
exports.ApolloQueryObservable = ApolloQueryObservable;
exports.CLIENT_MAP = CLIENT_MAP;
exports.CLIENT_MAP_WRAPPER = CLIENT_MAP_WRAPPER;
exports.ApolloModule = ApolloModule;
exports.APOLLO_PROVIDERS = APOLLO_PROVIDERS;
exports.APOLLO_DIRECTIVES = APOLLO_DIRECTIVES;
exports.defaultApolloClient = defaultApolloClient;

Object.defineProperty(exports, '__esModule', { value: true });

})));
